components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java

```diff
@@ -0,0 +1,401 @@
+import org.wso2.carbon.database.utils.jdbc.exceptions.TransactionException;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.sql.DataSource;
+
+/**
+ * This utility class used to call JDBC with lambda expressions is different from JdbcTemplate class as it supports the
+ * use of named prepared statement. This class simplifies the usage of JDBC and helps to avoid
+ * common errors.
+ * Using this class we can execute insert, select, update and delete queries by providing necessary parameters to the
+ * methods. Transaction support is also available with this implementation for the query executions. All the SQL
+ * operations performed by this class are logged at the debug level.
+ */
+public class NamedJdbcTemplate {
+
+    private static final Logger log = LoggerFactory.getLogger(NamedJdbcTemplate.class);
+    private String driverName;
+    private String productName;
+    private DataSource dataSource;
+
+    public NamedJdbcTemplate(DataSource dataSource) {
+
+        this.dataSource = dataSource;
+    }
+
+    /**
+     * Provides the transaction support for the JDBC executions of the template.
+     *
+     * @param callable the SQL query execution method call.
+     * @return result set of domain objects of required type.
+     */
+    public <T extends Object, E extends Exception> T withTransaction(NamedExecuteCallable<T> callable) throws
+            TransactionException, E {
+
+        TransactionEntry transactionEntry = TransactionManager.getTransactionEntry();
+        if (transactionEntry.getTransactionDepth() == 0) {
+
+            try {
+                Connection connection = dataSource.getConnection();
+                transactionEntry.setAutoCommitState(connection.getAutoCommit());
+                connection.setAutoCommit(false);
+                transactionEntry.setConnection(connection);
+            } catch (SQLException e) {
+                throw new TransactionException(JdbcConstants.ErrorCodes.ERROR_CODE_GETTING_CONNECTION_FAILURE.
+                        getErrorMessage(), JdbcConstants.ErrorCodes.ERROR_CODE_GETTING_CONNECTION_FAILURE.
+                        getErrorCode(), e);
+            }
+        }
+        transactionEntry.incrementTransactionDepth();
+        TransactionManager.setTransactionEntry(transactionEntry);
+        try {
+            T result = callable.get(new NamedTemplate<>());
+            transactionEntry.decrementTransactionDepth();
+            if (transactionEntry.getTransactionDepth() == 0) {
+                transactionEntry.getConnection().commit();
+            }
+            return result;
+        } catch (Throwable t) {
+            //We catch any exception and rollback the transaction.
+            logDebug("Could not commit the transaction.", t);
+            transactionEntry.decrementTransactionDepth();
```

Comment:
What if line https://github.com/wso2/carbon-utils/pull/15/files#diff-b121996948ed254f79b9d4ec1d54ddd4R85 throws an exception ?. It will be caught by this catch and will again try to decrement number which causes the number to be -1. So the roll back or any recovery will not take place.

Reply:
Same procedure is followed in https://github.com/wso2/carbon-utils/blob/master/components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/JdbcTemplate.java  
So should this be corrected?

Reply:
will it be ideal to check if transactionEntry.getTransactionDepth() != -1 before executing line https://github.com/wso2/carbon-utils/pull/15/files#diff-b121996948ed254f79b9d4ec1d54ddd4R91 ?

Reply:
will add a condition to check if transactionEntry.getTransactionDepth() > 0 before executing line https://github.com/wso2/carbon-utils/pull/15/files#diff-b121996948ed254f79b9d4ec1d54ddd4R91